graphics-items/instantiated-model-item.cpp

```diff
@@ -0,0 +1,161 @@
+//_/_/ CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+//_/_/ INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+//_/_/ MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+//_/_/ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
+//_/_/ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+//_/_/ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+//_/_/ BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+//_/_/ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+//_/_/ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+//_/_/ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+//_/_/ NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+//_/_/ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY 
+//_/_/ OF SUCH DAMAGE.
+//_/_/ 
+//_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/
+
+#include <regex>
+#include "aera-visualizer-window.hpp"
+#include "instantiated-model-item.hpp"
+#include <QMenu>
+#include "explanation-log-window.hpp"
+#include "aera-visualizer-scene.hpp"
+
+using namespace std;
+using namespace r_code;
+
+namespace aera_visualizer {
+
+ImdlItem::ImdlItem(
+  NewInstantiatedModelEvent* imdl, ReplicodeObjects& replicodeObjects,
+  AeraVisualizerScene* parent)
+  : AeraGraphicsItem(imdl,
+    replicodeObjects,
+    parent,
+    "Model " + makeHtmlLink(imdl->baseModel_, replicodeObjects) +
+    " " + RightDoubleArrowHtml + "<br>&nbsp;&nbsp;Instantiated Model "),
+  imdl_(imdl),
+  showState_(HIDE_IMDL),
+  replicodeObjects_(replicodeObjects)
+{
+  // Extract the 2nd set of bracketed valuesfrom an expression of the form
+  //    (imdl mdl_74 [10 0s:300ms:0us 0s:400ms:0us] [b v4: v5: v6: v7:] : :)
+  smatch matches;
+  regex templateValsRegex("\\[[\\w\\s\\.:-]*\\] (\\[[\\w\\s\\.:-]*\\])");
+  std::string sourceCode = replicodeObjects_.getSourceCode(imdl_->imdl_);
+  if (regex_search(sourceCode, matches, templateValsRegex)) {
+    templateVals_ = QString::fromStdString(matches[1].str());
+  }
+
+  // Set up the graphics item
+  setTextItemAndPolygon(makeHtml(), true);
+}
+
+
+QString ImdlItem::makeHtml()
+{
+  QString html;
+  QString imdlSource = QString::fromStdString(replicodeObjects_.getSourceCode(imdl_->imdl_));
+  QString baseMdlLabel = QString::fromStdString(replicodeObjects_.getLabel(imdl_->baseModel_));
+
+  if (showState_ == HIDE_IMDL) {
+    html += SelectedRadioButtonHtml + " Hide imdl" +
+      " <a href=\"#what-made-this\">" + UnselectedRadioButtonHtml + " What made this?</a><br>";
+
+    html += imdlSource.replace(baseMdlLabel, makeHtmlLink(imdl_->baseModel_, replicodeObjects_));
+
+  }
+  else if (showState_ == WHAT_MADE_THIS) {
+    html += "<a href=\"#hide-imdl\">" + UnselectedRadioButtonHtml + " Hide imdl</a>" +
+      " " + SelectedRadioButtonHtml + " What made this?<br>";
+    
+    html += "Input " + makeHtmlLink(imdl_->object_, replicodeObjects_) + " matched the LHS of model " +
```

Comment:
This prints, for example, "Input fact_406 matched the LHS..." But `imdl_->object_` is fact_406 which is already this instantiated model (not the input). We want to print something similar to `modelReduction_->getCause()` here:
https://github.com/IIIM-IS/AERA_Visualizer/blob/a06c43cee207e157593fc358caf8d2e5fe370920/graphics-items/prediction-item.cpp#L247

Notice that `ModelMkValPredictionReduction` stores the reduction object. This is because everything else (including the imdl and the input cause) can be obtained from it. Can you change `NewInstantiatedModelEvent` to store the reduction also? This way you can get the input cause from it.

Reply:
Okay, I see what you mean! I just updated the event to include the `getCause()` function from `ModelMkValPredictionReduction` and the imdl explanations seem to agree with the prediction explanations now. I also cleaned up the code a bit.